src/Simplex/Messaging/Agent/Transmission.hs

```diff
@@ -103,60 +107,82 @@ data SMPMessage
         previousMsgHash :: ByteString,
         agentMessage :: AMessage
       }
+  deriving (Show)
 
 data AMessage where
   HELLO :: VerificationKey -> AckMode -> AMessage
   REPLY :: SMPQueueInfo -> AMessage
   A_MSG :: MsgBody -> AMessage
+  deriving (Show)
 
 parseSMPMessage :: ByteString -> Either ErrorType SMPMessage
-parseSMPMessage _ = Left INTERNAL
+parseSMPMessage =
+  first (const $ SYNTAX errBadMessage)
+    . A.parseOnly (smpMessageP <* A.endOfLine <* A.endOfInput)
+  where
+    smpMessageP :: Parser SMPMessage
+    smpMessageP =
+      smpConfirmationP <* A.endOfLine
+        <|> A.endOfLine *> smpClientMessageP
+
+    smpConfirmationP :: Parser SMPMessage
+    smpConfirmationP = SMPConfirmation <$> ("KEY " *> base64P <* A.endOfLine)
+
+    smpClientMessageP :: Parser SMPMessage
+    smpClientMessageP =
+      SMPMessage
+        <$> A.decimal
+        <*> (A.space *> tsIso8601P)
+        <*> (A.space *> base64P)
+        <*> (A.endOfLine *> agentMessageP)
+
+    tsIso8601P :: Parser UTCTime
+    tsIso8601P = maybe (fail "timestamp") pure . parseISO8601 . B.unpack =<< A.takeTill (== ' ')
 
 serializeSMPMessage :: SMPMessage -> ByteString
 serializeSMPMessage = \case
-  SMPConfirmation sKey -> "KEY " <> sKey <> "\r\n\r\n"
+  SMPConfirmation sKey -> smpMessage ("KEY " <> encode sKey) "" ""
   SMPMessage {agentMsgId, agentTimestamp, previousMsgHash, agentMessage} ->
-    "\r\n" <> messageHeader agentMsgId agentTimestamp previousMsgHash <> "\r\n" <> serializeAgentMessage agentMessage
+    let header = messageHeader agentMsgId agentTimestamp previousMsgHash
+        body = serializeAgentMessage agentMessage
+     in smpMessage "" header body
+  where
+    messageHeader msgId ts prevMsgHash =
+      B.unwords [B.pack $ show msgId, B.pack $ formatISO8601Millis ts, encode prevMsgHash]
+    smpMessage smpHeader aHeader aBody = B.intercalate "\n" [smpHeader, aHeader, aBody, ""]
+
+agentMessageP :: Parser AMessage
+agentMessageP = hello <|> reply <|> a_msg
+  where
+    hello = "HELLO " *> (HELLO <$> base64P <*> ackMode)
+    reply = "REPLY " *> (REPLY <$> smpQueueInfoP)
+    a_msg = do
+      size <- "MSG " *> A.decimal :: Parser Int
+      A_MSG <$> (A.endOfLine *> A.take size <* A.endOfLine)
+    ackMode = " NO_ACK" $> AckMode Off <|> pure (AckMode On)
+
+smpQueueInfoP :: Parser SMPQueueInfo
+smpQueueInfoP =
+  SMPQueueInfo <$> ("smp::" *> smpServerP <* "::") <*> (base64P <* "::") <*> base64P
+
+smpServerP :: Parser SMPServer
+smpServerP = SMPServer <$> server <*> port <*> msgHash
   where
-    messageHeader agentMsgId agentTimestamp previousMsgHash =
-      B.unwords [B.pack $ show agentMsgId, B.pack (formatISO8601Millis agentTimestamp), encode previousMsgHash]
+    server = B.unpack <$> A.takeTill (A.inClass ":# ")
+    port = A.char ':' *> (Just . show <$> (A.decimal :: Parser Int)) <|> pure Nothing
+    msgHash = A.char '#' *> (Just <$> base64P) <|> pure Nothing
```

Comment:
what is the purpose of `msgHash` in smp server? is it for chain verification purposes?

Reply:
no, it's a pre-shared hash of the key (or some other secret) for transport encryption between server and the client - there should be some ad-hoc or noise-based protocol to negotiate a symmetric key for the TCP session.